toolsrc/include/vcpkg/base/jsonreader.h

```diff
@@ -0,0 +1,315 @@
+    {
+        switch (value.kind())
+        {
+            case ValueKind::Null: return visit_null(r);
+            case ValueKind::Boolean: return visit_boolean(r, value.boolean());
+            case ValueKind::Integer: return visit_integer(r, value.integer());
+            case ValueKind::Number: return visit_number(r, value.number());
+            case ValueKind::String: return visit_string(r, value.string());
+            case ValueKind::Array: return visit_array(r, value.array());
+            case ValueKind::Object:
+            {
+                const auto& obj = value.object();
+                r.check_for_unexpected_fields(obj, valid_fields(), type_name());
+                return visit_object(r, obj);
+            }
+            default: vcpkg::Checks::unreachable(VCPKG_LINE_INFO);
+        }
+    }
+
+    template<class Type>
+    Optional<Type> IDeserializer<Type>::visit(Reader& r, const Object& obj)
+    {
+        r.check_for_unexpected_fields(obj, valid_fields(), type_name());
+        return visit_object(r, obj);
+    }
+
+    template<class Type>
+    View<StringView> IDeserializer<Type>::valid_fields() const
+    {
+        return {};
+    }
+
+    template<class Type>
+    Optional<Type> IDeserializer<Type>::visit_null(Reader&)
+    {
+        return nullopt;
+    }
+    template<class Type>
+    Optional<Type> IDeserializer<Type>::visit_boolean(Reader&, bool)
+    {
+        return nullopt;
+    }
+    template<class Type>
+    Optional<Type> IDeserializer<Type>::visit_integer(Reader& r, int64_t i)
+    {
+        return this->visit_number(r, static_cast<double>(i));
+    }
+    template<class Type>
+    Optional<Type> IDeserializer<Type>::visit_number(Reader&, double)
+    {
+        return nullopt;
+    }
+    template<class Type>
+    Optional<Type> IDeserializer<Type>::visit_string(Reader&, StringView)
+    {
+        return nullopt;
+    }
+    template<class Type>
+    Optional<Type> IDeserializer<Type>::visit_array(Reader&, const Array&)
+    {
+        return nullopt;
+    }
+    template<class Type>
+    Optional<Type> IDeserializer<Type>::visit_object(Reader&, const Object&)
+    {
+        return nullopt;
+    }
+
+    VCPKG_MSVC_WARNING(pop);
+
+    struct StringDeserializer final : IDeserializer<std::string>
+    {
+        virtual StringView type_name() const override { return type_name_; }
+        virtual Optional<std::string> visit_string(Reader&, StringView sv) override { return sv.to_string(); }
+
+        explicit StringDeserializer(StringLiteral type_name_) : type_name_(type_name_) { }
+
+    private:
+        StringLiteral type_name_;
+    };
+
+    struct PathDeserializer final : IDeserializer<fs::path>
+    {
+        virtual StringView type_name() const override { return "a path"; }
+        virtual Optional<fs::path> visit_string(Reader&, StringView sv) override { return fs::u8path(sv); }
+
+        static PathDeserializer instance;
```

Comment:
```suggestion
        static const PathDeserializer instance;
```

Could we do this, and also make all the `visit` functions `const`? There's no reason for them to be non-`const`.

(same for the rest of the instances)

Reply:
This is true for now, however I didn't want to assume that all deserializers will forever not need mutable members.

Regardless, I think if we want to make this change it can be made in a separate PR.

Reply:
Global mutable state is an extreme code smell, though... even though it's fine here, I don't like that we have it at all.